### ISSUE TITLE
GRU?

### ISSUE THREAD

**Issue by CCorfield** [SeanNaren/CTCNetworkClient#7]
What would it take to add a GRU variation to compare with LSTMs?

**Comment by SeanNaren** [comment]
Sounds reasonable, I know in the Deep speech paper they do this so I don't see why not! 

The current branch now supports CuDNN RNN layers. I'll add an option to switch to cudnn Bi-directional GRUs ASAP.

**Comment by SeanNaren** [comment]
Added a GRU option to the main branch. In order to use it you must set the option in the AN4CTCTrain to true. Let me know if this is what you were after :)

**Comment by CCorfield** [comment]
Before I wrote the message, I had tried to add a GRU layer, without much luck (no support for bxtxf). By fortunate coincidence  you had just upgraded to the latest cudnn, which has the GRU with batch support, however you may want to include a note in the set-up docs that the luarocks install for Soumith's R5 torch bindings doesn't bring down everything, and you may have to manually pull down all the files and install in your cudnn folder in torch installation (pending a fix in the R5 repo). On a different note, the academic in me thinks that the output layer should really be SoftMax or LogSoftMax.Anyhow, my compliments to the chef. -CC

**Comment by SeanNaren** [comment]
Yeah I've noted install the bindings for R5 is a bit difficult, not sure why it is like that or if I'm handling it wrong since you could just add branch = 'R5' in the rockspec.

You are right the last layer should be a SoftMax however from what I've seen in the source code (and the major issues I ran into) this is done internally in warp-ctc!

**Comment by CCorfield** [comment]
AN4CTCTest.lua: Don't forget to add "GRU" to the network parameters :-)

Regarding SoftMax, it would be interesting to see if it can be pulled into the main network, because predictions from a trained network won't have the benefit of the CTC criterion being called. I have also wondered if it would make sense to do a batch normalization before the spatial convolution. The reason being that legacy reco systems (using MMFCC's) will do a "CMV" (Cepstral Mean and Variance ) right up front. The spatial convolutions should be able to do the equivalent of "VTLN" (Vocal Tract Length Normalization) in addition to feature spotting. And, lastly, Baidu went for ReLU, but I wonder if a  HardTanh() would be better -- just as quick to calculate, but keeps the data in the range (-1,1) which might be better for a final SoftMax layer.  Some fun things to try out here in the boonies. Btw: Where are you based?

**Comment by SeanNaren** [comment]
@CCorfield Wow these all sound like great feedback thanks! A lot of terms being thrown around here that I'm gonna have to read up on :P

I agree never thought of it that way, would be a bit inconvenient receiving the activations and them not being bound to something. Sadly I'm not sure how easy this would be without having to modify the source C code for warp-ctc!

I've made some changes to the master branch that have led to results getting a bit iffy. Once I've fixed this up I'll modify the code so that it is easy to plug and play different net architectures to test them individually. Would make testing improvements with HardTanh() and batch norm easier! 

A quick question, I've been having an issue with adding batch norm to the bi-recurrent layers in that I don't know the appropriate place to add them. How would you suggest approaching this?

Currently based in UK London, this was my undergrad final year project, but it is far from done :)

**Comment by CCorfield** [comment]
A few pointers regarding the acronyms: Historically, the speech reco folks have processed the audio into "feature vectors" consisting of "Mel (Frequency) Cepstral Coefficients" (MFCCs). Trial (and much error) suggested that they contained the real information in a denser format than straight spectrograms. The Baidu folks might argue that we have a lot more computing power today and we can let a neural net figure out where the information is buried. That being said, it would be interesting to do an apples to apples comparison and take the same network but train it on MFCCs instead of the spectrogram. If you look up MFCCs, don't be surprised if they seem a little weird and on their way to the dustbin, much as astronomers no longer bother to learn spherical geometry, since computers do it all nowadays. 
The CMV normalization is exactly the analog of various types of Normalization routine you see in Torch -- the idea was to massage the data to mean = 0, and std_dev = 1. Historically, in speech reco it was (and still is) applied up front. VLTN is a fancy way of observing that people's voices have different pitches and harmonics, because the resonant frequencies of their vocal tracts depend on the dimensions of their vocal tracts. However, if you look at spectrograms in "log frequency" space, it amounts to a linear shift up or down. This is why the networks use spatial convolutions: they are good at spotting features which "look" the same but are displaced linearly from each other. The reference to "log frequency" spectrum reminds me to explain "Mel" frequencies: human perception of differences in pitch is roughly logarithmic, but the best empirical approximation is to turn real frequencies into Mel frequencies (you can look up the formula online), and use these as the basis for pitch differences in speech. The "Cepstrum" bit, is simple taking the distribution of energy in various bands of Mel frequency space. You could stop there -- an energy sampling in Mel space -- and feed those into the neural network. However, the powers-that-be in speech reco always apply a discrete cosine transformation (like a Fourier Transform) first. Hopefully, this comment was too much of mouthful. Cheers, CC

**Comment by SeanNaren** [comment]
Awesome stuff thanks for the explanation! I'll try to experiment as much as I can with the different configurations. 

Not sure if you saw my last point (bit off topic) about the batch norm, any ideas as to how batch norm could be applied in a straight forward manner to the bi-directional layers?

**Comment by CCorfield** [comment]
Sorry -- I missed your question about batch norm and bi-directional layers. Just before giving a quick -- and totally wrong(!) -- reply, let me experiment. It reminds me of an incident in the life the great statistician Fisher, who, in the middle of a lecture, casually remarked "It is obvious that..." And paused. And left the lecture theater. And, ten minutes later, he re-entered and announced "Yes, it is obvious." and carried on giving the lecture. Some people have style.

**Comment by CCorfield** [comment]
I saw you have been busy on the project in the last 24 hours. And it looks like you answered your question about the number of dimensions coming through the layers and their consequent impact on batch normalization. I pulled down a clean copy and can confirm to you that it installed and ran fine. 

Since you are (in a very good sense!) a maniac for perfection, Allow me to suggest you include running a validation set through the net at the end of every training epoch, so that you can plot a graph of how the WER evolves on held-out data. One of the issues with nets is over-fitting where they become too good at handling the training data, and worse at handling novel inputs. Running validation data through the net supplies a stopping criterion -- the WER on the validation set will drop to a minimum and then may start to rise again, in which case your "best" model is the one at the minimum. So, you can save snapshots with idea of that the one corresponding to the validation set's minimum WER is your best net. Hopefully, it goes without saying, that you want to run the held-out data through the net without it tainting the training (i.e., so that there won't be any accumulated values used in the next backward/update step).

Btw -- I was wondering what system you are doing your development on -- a home system (Linux box with a GPU card, or two) or are you doing it on your university's computer system? You are making great progress and it is a joy to watch your project unfold from the 'peanut gallery'. Cheers, -CC

**Comment by SeanNaren** [comment]
hahaha thanks for your aid so far! Been awesome :) Definitely not a perfectionist but am trying to get the best out of the stuff!

Really good idea, will definitely implement the validation set. Not sure what you mean by accumulated values in the next backwards call though, wouldn't the validation be separate to the training process?

Yep I'm using my local computer with a GTX 970 with 4GB VRAM and an I5 ivy bridge processor running Linux 14.04. I should appreciate what I have but extra vram wouldn't hurt since it will take a ridiculously long time to train on the voxforge dataset with such small batch sizes... What have you been running on?

Thanks for your kind words, means a lot! On quick thing to mention is on the topic of batch norm I just kinda put them in the places that made some sense, since layer wise batch norm isn't possible with the cuDNN library unless I manually do the layers myself (which removes all the parallelisation that they did with their library).

**Comment by CCorfield** [comment]
Regarding intermixing training and evaluation: at the end of an epoch you could do something like:
  model:evaluate()
  model:forward(validationset)
  calculate and send the WER to a log
  model:train()
And then go and do the next epoch. 

I have a Linux laptop with the same GPU as you (it has two installed). There's a local company which builds and sells Linux/Ubuntu systems. The term 'laptop' does not do justice to the weight/size of the beastie, a better term would "thudtop". It probably has lead weights inside.

**Comment by SeanNaren** [comment]
Awesome! Will get it implemented :) 

A laptop with two GTX 970s doesn't sound very lap friendly...

**Comment by CCorfield** [comment]
In terms of refining what the layers do, here are a couple of thoughts:
1. I would do a SpatialBatchNormalization on entry into the network. The broadest brush stroke approach is to normalize across the entire batch, but another view is that it should be done for each batch member individually. In the real world, the audio level for any speaker will vary by how loudly they are speaking and by the distance to the microphone. There can be variation between speech segments/samples for the same speaker. If you normalize right up front then the network will always see data with same mean and variance. The subsequent normalizations mitigate the spurious effects of changing the values of coefficients in the matrices W & B where the changes amplify and shift the output without adding any meaningful information (socalled covariant drift). 
2. Bidirectional layers are potentially problematic, because they process from both ends of a sample, which means they have an advantage that they "know" what is going to happen in the future. This is OK if you want to get the highest accuracy for offline or batch transcription, but if you want to mimic human speech processing then you can't "look ahead". This suggests that regular LSTM / GRU recurrent layers would be the preferred implementation for real time speech processing. In DS2, the single directional layers compared favorably with the bidirectional ones. 
3. In your latest network you appear to transposed batch and time so that you are putting t x b x f through the recurrent layer(s). Are you sure that you want to do it that way and not b x t x f?

When AN4 is up and running with the kinks sorted out and you want to run a bigger corpus through a system with more capacity than your home computer let me know and I can help out on that front. 

Cheers, -CC

**Comment by SeanNaren** [comment]
Thanks for your feedback, I've done some changes locally to bring back normalization to the network and have put them in the correct places, will add soon. 

On the topic of bi-directional, completely correct it would not work for real time speech and definitely worth checking out. I'll check it out ASAP and report back! 

Finally yep it has to be `seqLength x batch x features` for the cuDNN library. I'm not using the optional variable `batchFirst` because it has a few issues that I've got a PR open for. Again, thanks for all your help and offer!
